File: algorithm/40. Combination Sum II/40.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        result = []
        if not candidates:
            return result
        candidates.sort()
        self.helper(candidates, target, [], 0, result)
        return result

    def helper(self, candidates, target, path, index, result):
        if target == 0:
            result.append(path[:])
            return

        for i in range(index, len(candidates)):
            if i!=index and candidates[i]==candidates[i-1]:
                continue
            if candidates[i] > target:
                break
            self.helper(candidates, target - candidates[i], path + [candidates[i]], i+1, result)

File: algorithm/40. Combination Sum II/40.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        Arrays.sort(candidates);
        helper(candidates, target, res, new ArrayList<Integer>(), 0);
        return res;
    }
    
    public void helper(int[] candidates, int target, List<List<Integer>> res, List<Integer> com, int index)
    {
        if(target == 0)
        {
            res.add(new ArrayList<Integer>(com));
            return;
        }
        for(int i = index; i < candidates.length && target >= candidates[i]; i++)
        {
            if(i == index || candidates[i] != candidates[i - 1])
            {
                com.add(candidates[i]);
                helper(candidates, target - candidates[i], res, com,i + 1);
                com.remove(com.size() - 1);   
            }
        }
        return;
    }
}
